Doc: Marshal.to_channel may throw exception

kkirstein opened the pull request:
`Marshal.to_channel` throws an exception if the provided channel is not in binary mode (mostly an issue on Windows). This should be mentioned in the docs.
Here is the respective snippet of `caml_output_val` (called by `Marshal.to_channel` via `caml_output_value`):
```
void caml_output_val(struct channel *chan, value v, value flags)
{
  char header[32];
  int header_len;
  struct output_block * blk, * nextblk;

  if (! caml_channel_binary_mode(chan))
    caml_failwith("output_value: not a binary channel");
[..]
```

kirk-ut committed:
Doc: Marshal.to_channel may throw exception

`Marshal.to_channel` throws an exception if the provided channel is not in binary mode (mostly an issue on Windows). This should be mentioned in the docs.
Here is the respective snippet of `caml_output_val` (called by `Marshal.to_channel` via `caml_output_value`):
```
void caml_output_val(struct channel *chan, value v, value flags)
{
  char header[32];
  int header_len;
  struct output_block * blk, * nextblk;

  if (! caml_channel_binary_mode(chan))
    caml_failwith("output_value: not a binary channel");
[..]
```

nojb commented on stdlib/marshal.mli:
```diff
@@ -109,6 +109,9 @@ val to_channel : out_channel -> 'a -> extern_flags list -> unit
    only matters when marshaling is performed on a 64-bit platform;
    it has no effect if marshaling is performed on a 32-bit platform.
+   
+   It raises [Failure "output_value: not a binary channel"] if channel
+   [chan] is not in binary mode.
```
```suggestion
   It raises [Failure _] if [chan] is not in binary mode.
```
It is better not to depend on the exact error message.

dra27 replied:
Some of the documentation precedes ocamldoc having `@raise`, but for new changes it's nice to use it, even if other functions aren't (yet) updated:

```suggestion
   @raise Failure if [chan] is not in binary mode.
```

nojb reviewed (commented):
LGTM, modulo a small observation.

gasche commented on stdlib/marshal.mli:
```diff
@@ -109,6 +109,9 @@ val to_channel : out_channel -> 'a -> extern_flags list -> unit
    it has no effect if marshaling is performed on a 32-bit platform.
+   
+   It raises [Failure "output_value: not a binary channel"] if channel
+   [chan] is not in binary mode.
```
The consensus is that we should not be explicit about the string message in the specification of the function, to be allowed to change it in the future. So maybe just say that a `Failure` exception is raised?

kkirstein replied:
Agreed! I took the doc from `Marshal.from_channel` as blue print, so we might omit the message string there as well. Do you prefer `Failure` or `Failure _` as suggested by @nojb ?

kirk-ut committed:
use @raise tag for docs

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

kirk-ut committed:
update doc with new @raise tag

As suggested for `Marshal.to_channel` update doc for `Marshal.from_channel` to omit message string and use @raise tag.

xavierleroy commented on stdlib/marshal.mli:
```diff
@@ -109,6 +109,8 @@ val to_channel : out_channel -> 'a -> extern_flags list -> unit
    when read back on a 32-bit platform.  The [Mashal.Compat_32] flag
    only matters when marshaling is performed on a 64-bit platform;
    it has no effect if marshaling is performed on a 32-bit platform.
+   
+   @raise Failure if [chan] is not in binary mode.
```
Why not add the same info to `Marshal.from_channel`?  It also fails if the  channel is not in binary mode.

kkirstein replied:
Sure, I must have overlooked that.

xavierleroy replied:
Would you please add the info as I suggested, instead of resolving this conversation and doing nothing?

kkirstein replied:
Sorry, my comment above was misleading (.. and I am quite a newbie with this PR review process). I already did that change in my branch, so it should be included in this PR (see commit [e6d11e0](https://github.com/ocaml/ocaml/pull/10012/commits/e6d11e02c4747a5e56581743051c613ed4251ca4) ).

xavierleroy replied:
Sorry for the possible misunderstanding, but I just downloaded your branch, and in `stdlib/marshal.mli` I see:
```
   @raise Failure if [chan] is not in binary mode.
```
in the documentation comment for `Marshal.to_channel`, but I do not see it in the documentation comment for `Marshal.from_channel`, which I copy verbatim below for reference:
```
val from_channel : in_channel -> 'a
(** [Marshal.from_channel chan] reads from channel [chan] the
   byte representation of a structured value, as produced by
   one of the [Marshal.to_*] functions, and reconstructs and
   returns the corresponding value.

   @raise End_of_file if [chan] is already at the end of the file.

   @raise Failure if the end of the file is reached during
   unmarshalling itself. *)
```
So, would you please add `@raise Failure if [chan] is not in binary mode.` to the documentation comment for `Marshal.from_channel` as I already requested TWICE?  Thank you.

kkirstein replied:
Oops, now I got it. Sorry again, I was referring to the quoted line numbers instead of looking up the cited function. Should be _really_ added now. I did some rewording, as `Marshal.form_channel` might also throw `Failure` for another reason.

xavierleroy reviewed (commented):
One suggestion below.

dra27 commented on stdlib/marshal.mli:
```diff
@@ -109,6 +109,8 @@ val to_channel : out_channel -> 'a -> extern_flags list -> unit
    when read back on a 32-bit platform.  The [Mashal.Compat_32] flag
    only matters when marshaling is performed on a 64-bit platform;
    it has no effect if marshaling is performed on a 32-bit platform.
+   
```
```suggestion

```

dra27 commented on stdlib/marshal.mli:
```diff
@@ -139,10 +141,11 @@ val from_channel : in_channel -> 'a
-   end of file when starting to read from the channel, and raises
-   [Failure "input_value: truncated object"] if it reaches the end
-   of file later during the unmarshalling. *)
+   @raise End_of_file if the function has already reached the end
+   of file when starting to read from [chan].
+   
+   @raise Failure if it reaches the end of the file during
+   the unmarshalling. *)
```
```suggestion
   @raise End_of_file if [chan] is already at the end of the file.

   @raise Failure if the end of the file is reached during
   unmarshalling itself. *)
```
I suggest the re-wording only because the whitespace needed updating anyway - I think it makes the point more clearly that if you read a value from a file which is at EOF then you'll get `End_of_file` (i.e. you can read a sequence of values of from a file in the same way as you'd read a sequence of lines, terminating when you encounter `End_of_file`) but if you get `Failure` if the function reaches EOF _while_ reading a value.

dra27 reviewed (commented):
Thanks!

There's stray whitespace which or check-typo script is catching - as that has to be fixed, I propose an alteration to the wording of the two `@raise` lines for `from_channel` since effectively splitting the original sentence loses part of the explanation

kirk-ut committed:
Update stdlib/marshal.mli

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

kirk-ut committed:
Update stdlib/marshal.mli

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

kirk-ut committed:
add exception comment to Marshal.from_channel

dra27 reviewed (approved):
I confess managed to read @xavierleroy's request correctly, either. This looks good to merge to me, thank you!

xavierleroy reviewed (approved):
Looks good to me.  Apologies if my request for changes was not 100% clear to you, but in this case it's better to ask for clarifications rather than ignore the request.
